packages/toolpad-app/src/toolpad/AppEditor/AppEditorShell.tsx

```diff
@@ -111,18 +123,52 @@ export default function AppEditorShell({ appId, ...props }: ToolpadAppShellProps
   const domLoader = useDomLoader();
 
   const [createReleaseDialogOpen, setCreateReleaseDialogOpen] = React.useState(false);
+  const [isSaveStateVisible, setIsSaveStateVisible] = React.useState(false);
+
+  const hasUnsavedChanges = domLoader.unsavedChanges > 0;
+  const isSaving = domLoader.saving;
+
+  React.useEffect(() => {
+    let timeout: NodeJS.Timeout | null = null;
+
+    if (!hasUnsavedChanges) {
+      timeout = setTimeout(() => {
+        setIsSaveStateVisible(false);
+
+        if (timeout) {
+          clearTimeout(timeout);
+        }
+      }, 4500);
+    } else {
+      setIsSaveStateVisible(true);
+    }
+
+    return () => {
+      if (timeout) {
+        clearTimeout(timeout);
+      }
+    };
+  }, [hasUnsavedChanges]);
 
   return (
     <ToolpadAppShell
       appId={appId}
       actions={
         <Stack direction="row" gap={1} alignItems="center">
-          {domLoader.saving ? (
-            <Box display="flex" flexDirection="row" alignItems="center">
-              <CircularProgress size={16} color="inherit" sx={{ mr: 1 }} />
-            </Box>
+          {isSaveStateVisible ? (
+            <Tooltip title={getSaveStateMessage(isSaving, hasUnsavedChanges)}>
+              <Box display="flex" flexDirection="row" alignItems="center">
+                {isSaving ? (
+                  <CircularProgress size={16} color="inherit" sx={{ mr: 1 }} />
```

Comment:
I feel that we don't need the `CircularProgress` icon for the `isSaving` state: it is not discernable to me in the video as a separate icon; perhaps we can add a "pulsating" animation to the icon in the `isSaving` state

Reply:
Anyway, that could be a follow-up improvement! This looks great to me overall as a first step

Reply:
>it is not discernable to me in the video as a separate icon

keep in mind this video is made on a locally running app, irl it will often take longer to complete the request

Reply:
yeah it's showing as long as `saving` is true, looks like in dev it's a pretty short time.
lmk if you think we should show the loader when there are unsaved changes even though the saving flag is `false`